### ISSUE TITLE
[wishlist] "multistage" benchmarks via Python generators

### ISSUE THREAD

**Issue by yarikoptic** [airspeed-velocity/asv#904]
TL;DR proposal:  support "generator" benchmarks, where asv would time it took from one stage in the benchmark to the next (or completion).  Stages could be easily exposed by a benchmark by yielding after each stage completion, e.g.

```python
def time_life_cycle1():
     ds = create_dataset()
     yield "create"

     perform_computation()
     yield  #  yielding None would indicate that time from "create" to this point is ignored

     assert ds.is_dirty()
     yield "is_dirty on dirty"  # We are timing "is_dirty" command performance within this setup

     ds.save()
     yield "save"
     
     assert not ds.is_dirty()
     yield "is_dirty on clean"
```
in effect producing 4 benchmarks (could be named as e.g. `time_life_cycle1::is_dirty on clean` etc), by timing durations from initial invocation to any next `yield` (ignoring the ones yielding None, just resetting the timer on those).  That would efficiently eliminate need for complex `setup()` procedures to arrive to the specific state within such a "life cycle" (or pipeline etc).

Aforementioned example is for a prototypical benchmark for DataLad but I would love to use something like for other projects, e.g. while decoupling basic IO on the file from Python-side manipulation of just loaded basic data structures (by yielding right after underlying IO operation, and proceeding with stages in Python).

It could also make it easier to write `setup`/`time_`/`teardown`  within a single `time_` benchmark (by yielding None after setup block, and after teardown) where it is a single benchmark for setup/teardown.

I guess the difficulty in implementation could come from the fact that full "name" of the benchmark would not be known until it is ran.  But I guess situation is similar to the parametric benchmarks.

Please let me know if you feel that it might be a welcomed contribution to asv, so may be I could try to provide at least an initial implementation.
